js/binary/encoder.js

```diff
@@ -409,19 +409,34 @@ jspb.BinaryEncoder.prototype.writeFixedHash64 = function(hash) {
  */
 jspb.BinaryEncoder.prototype.writeString = function(value) {
   var oldLength = this.buffer_.length;
-
-  // UTF16 to UTF8 conversion loop swiped from goog.crypt.stringToUtf8ByteArray.
+ 
   for (var i = 0; i < value.length; i++) {
+    
     var c = value.charCodeAt(i);
+    // Look for surrogates
+    if (c >= 0xD800 && c <= 0xDBFF && i + 1 < value.length) {
```

Comment:
Please put this check inside the branch `else if (c < 65536)`. There is no reason to check this in the `c < 128` and `c < 2048` cases.